xdsl/traits.py

```diff
@@ -23,7 +24,33 @@ class OpTrait:
     Note that traits are the merge of traits and interfaces in MLIR.
     """
 
-    parameters: Any = field(default=None)
+    _parameters: Any = field(default=None)
+
+    @property
+    def parameters(self) -> Any:
+        return self._parameters
+
+    def verify(self, op: Operation) -> None:
+        """Check that the operation satisfies the trait requirements."""
+        pass
+
+
+@dataclass(frozen=True)
+class LazyOpTrait(OpTrait):
+    """
+    A lazy version of OpTrait which postpone the evaluation of parameters.
+    It can used to work around the circular dependencies in code.
+    """
+
+    _parameters: Callable[[], Any] | Any = field(default=None)
```

Comment:
There are no checks here that `_parameters` are callable when creating the object.

Reply:
`TypeError` will be thrown if it is not callable. Or you want to have a more specific error message/check? Also current implementation is lazily checking the type of the parameter given.